proposals/p3646.md

```diff
@@ -0,0 +1,252 @@
+
+## Proposal
+
+Formally, we have not yet approved a proposal that says that Carbon has tuple
+types, although we have approved several proposals that explicitly include
+support for tuples. So, this proposal does that: tuples exist in Carbon, and are
+product types with unnamed positional elements.
+
+This proposal also updates the design to match other decisions that have been
+made in leads issues but not captured by a proposal, specifically:
+
+-   Leads issue #2191 (one-tuples and one-tuple syntax), despite being focused
+    on one-tuples, established the syntax for tuples of all arities.
+-   Leads issue #710 established rules for assignment, comparison, and implicit
+    conversion of tuples. These operations are performed elementwise, with
+    relational comparisons being performed lexicographically.
+
+Finally, the main intent of this proposal is to add support for indexing tuples,
+using the following syntaxes:
+
+-   `.` _N_, where _N_ is an integer literal, and
+-   `.` `(` _expr_ `)`, where _expr_ is a template constant of integer type.
+
+## Details
+
+### Lexing
+
+Multi-level tuple indexing will result in constructs such as
+`tuple_of_tuples.1.2`. It's important that these are lexed as two tuple indexing
+operations, not as `tuple_of_tuples` `.` `1.2`, as it would be under the current
+lexical rules, so a new rule is introduced:
+
+-   When a `.` or `->` token is followed immediately by a digit, it is lexed as
+    a `.` or `->` token followed by an integer literal, never a real literal.
+
+Note that this results in lexing being slightly contextual: the rule to lex a
+token after a `.` or `->` is different from the rule to lex a token in any other
+context. However, there is an alternative equivalent formulation of the rule
+that is not context-sensitive: that `.integer` is treated as a single lexeme
+that produces two tokens, and likewise for `->integer`.
+
+### Precedence
+
+The `.` _N_ syntax has the same precedence as postfix member access syntax, `.`
+_name_, and can be combined in the same expression: `a.0.x.1` is valid.
+
+The `.` `(` _expr_ `)` syntax is not new in this proposal, and continues to have
+the same precedence as `.` _name_.
+
+### Expression operand
+
+In the `.` `(` _expr_ `)` syntax, if the first operand is a tuple and the second
+operand is a constant of any integer type, the result is the corresponding tuple
+element. This rule is built into the language; the `.` `(` ... `)` notion is not
+currently overloadable.
+
+### Bounds
+
+If the tuple index is not between 0 and one less than the number of elements in
+the tuple, inclusive, the indexing is invalid.
+
+### Tuple slicing
+
+The current skeleton design suggests using `tuple[a .. b]` to slice tuples. For
+example, `tuple[0 .. 2]` could be used to extract the first two elements of a
+tuple. Tuple slicing support is not covered by this proposal, but could be added
+in the future with syntax such as `tuple.(0 .. 2)`.
```

Comment:
I like this approach. I think it is worth documenting the concern that it might lead to an incorrect theory about how Carbon works, where `tuple.__` gives you a value but `tuple.(__)` gives you a tuple, though I don't think that should stop us from adopting this approach.

Reply:
Added some text mentioning that concern.